### PR TITLE
[Application] Support filtering apps by delegatee gateway address

### PR DESCRIPTION
**Pull request by adshmh**
## Summary

Support querying applications based on the delegated gateway address:
- Add a new field to the `QueryAllApplicationsRequest` type.
- Update the application keeper to process the query request based on the updated struct.
- Add unit tests to cover the new functionality.

Note: #888 adds a similar functionality by allowing access to delegating apps directly from a gateway struct. This PR  allows querying apps directly for a specific gateway delegation. Likely both PRs are needed in the long term.

## Issue

This is needed ot support integration with PATH. More specifically, it allows PATH to get the list of all applications which delegate to a specific gateway address:

https://github.com/buildwithgrove/path/blob/8533ee8b6dae4dcc5d4615bdd1b58d6229ef85a7/protocol/shannon/gateway_mode.go#L25-L26

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: x/application/types/query_validation.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// ValidateBasic performs basic (non-state-dependant) validation on a QueryAllApplicationsRequest.
+func (query *QueryAllApplicationsRequest) ValidateBasic() error {
+	delegateeGatewayAddr := query.GetDelegateeGatewayAddress()
+	if delegateeGatewayAddr == "" {
+		return nil
+	}
+
+	// Validate the delegation gateway address if the request specifies it as a constraint.
+	if _, err := sdk.AccAddressFromBech32(delegateeGatewayAddr); err != nil {
+		return ErrAppInvalidGatewayAddress.Wrapf("invalid gateway address for query: %q; (%v)", delegateeGatewayAddr, err)
```
Comment (red-0ne) [inline x/application/types/query_validation.go:16]:
```suggestion
		return ErrAppInvalidGatewayAddress.Wrapf("querying for apps with delegatee gateway address: %q; (%v)", delegateeGatewayAddr, err)
```

File: x/application/keeper/application_test.go [hunk 2]
```diff
@@ -18,18 +19,35 @@ import (
...
+		}
+
 		keeper.SetApplication(ctx, apps[i])
 	}
 	return apps
 }
 
+// testAppModifierDelegateeAddr adds the supplied gateway address to the application's delegatee list if the application's address matches
+// the supplied address list.
+func testAppModifierDelegateeAddr(delegateeGatewayAddr string, appsWithDelegationAddr []string) testAppModifier {
```
Comment (red-0ne) [inline x/application/keeper/application_test.go:43]: Wdyt about naming it `withAppDelegateeGatewayAddr`?

File: x/application/keeper/query_application_test.go [hunk 3]
```diff
@@ -122,3 +124,44 @@ func TestApplicationQueryPaginated(t *testing.T) {
...
+		addrInvalid := "invalid-address"
+		_, err := keeper.AllApplications(ctx, requestBuilder(addrInvalid))
+		expectedErr := status.Error(
+			codes.InvalidArgument,
+			types.ErrAppInvalidGatewayAddress.Error(),
+		)
+
+		require.ErrorContains(t, err, expectedErr.Error())
+	})
+}
```
Comment (red-0ne) [inline x/application/keeper/query_application_test.go:167]: It's worth adding a test with an empty address and assert that all apps are returned.

**Review by red-0ne — APPROVED** [review]
Thanks for putting this together.

Left a few comments but nothing blocking.

Also, could you please add the `devnet-test-e2e` label before pushing new changes to trigger e2e tests?

File: proto/poktroll/application/query.proto [hunk 4]
```diff
@@ -53,6 +53,8 @@ message QueryGetApplicationResponse {
 
 message QueryAllApplicationsRequest {
   cosmos.base.query.v1beta1.PageRequest pagination = 1;
+  // delegatee_gateway_address, if specified, filters the application list to only include those with delegation to the specified gateway address.
+  string delegatee_gateway_address = 2;
```
Comment (Olshansk) [inline proto/poktroll/application/query.proto:57]: Thoughts on renaming this to `gateway_address_delegated_to`.

I realize that `delegatee_gateway_address` more dramatically correct (probably?) but it confuses me when I read it and forces me to think.

When I read `gateway_address_delegated_to`, I don't need to think.

File: x/application/keeper/query_application_test.go (hunk 3 above)
Comment (Olshansk) [reply x/application/keeper/query_application_test.go:167]: +1

**Review by Olshansk — COMMENTED** [review]
@adshmh Leaving a comment (not a request for changes, nor an approval) because I'm curious to hear what you think of my feedback.

File: proto/poktroll/application/query.proto (hunk 4 above)
Comment (Olshansk) [reply proto/poktroll/application/query.proto:57]: 100% agreed.

TODO_MAINNET instead of beta though ;)

Comment (adshmh) [reply proto/poktroll/application/query.proto:57]: Done, thank you for the reviews.

File: x/application/keeper/application_test.go (hunk 2 above)
Comment (adshmh) [reply x/application/keeper/application_test.go:43]: Done, thank you for the review.

**Review by Olshansk — APPROVED** [review]
